### PR TITLE
[cli]: Expand `onboard` command with an entity discovery feature

### PR DESCRIPTION
**Pull request by marcuseide**
When running a fresh new Backstage installation the `catalog` is populated by a few example entities. We wanted to make it possible to create entities from a users actual code repositories instead - as we believe it helps a user to better understand the value Backstage can provide.

The `onboard` command can "discover" entities by analyzing code repositories. In this first iteration only basic analysis is supported, but could be expanded in the future. Those entities are then included in the `catalog` when running Backstage locally.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+          object(expression: "HEAD:") {
+            ... on Tree {
+              entries {
+                name
+                type
+                object {
+                  ... on Blob {
+                    byteSize
+                    text
+                    isBinary
```
Comment (mitchhentgesspotify) [inline packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts:72]: image

Are we fetching every file in a repository, _including its contents_?
I'm worried about performance here, especially since the only usage of this data is to get the contents of `package.json`.

Perhaps we should replace this function with something like:
```
async fetchPackageJson: Promise<text | undefined> {
    // fetch package.json, returning "undefined" if it doesn't exist
    // throw an error if a request fails for another reason, etc
    return unparsedJson
}
```

File: packages/cli/src/commands/onboard/discovery/providers/github/graphql.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+  transformer: (
+    item: GraphqlType,
+    ctx: TransformerContext,
+  ) => Promise<OutputType | undefined>,
+  variables: Variables,
+): Promise<OutputType[]> {
+  const result: OutputType[] = [];
+
+  let cursor: string | undefined = undefined;
+  for (let j = 0; j < 1000 /* just for sanity */; ++j) {
```
Comment (mitchhentgesspotify) [inline packages/cli/src/commands/onboard/discovery/providers/github/graphql.ts:52]: If we hit `1000`, this will fail silently with missing information.
I'd lean towards `while (hasNextPage)` instead, because at least if it hangs, we'll definitely get a bug report 👍

File: packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      const credentials = await this.#credentialsProvider.getCredentials({
+        url,
+      });
+      if (credentials.headers) {
+        return credentials.headers;
+      } else if (credentials.token) {
+        return { authorization: `Bearer ${credentials.token}` };
+      }
+    } catch {
+      // ignore silently
```
Comment (mitchhentgesspotify) [inline packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts:98]: If we have a bug in, say, `getCredentials()`, it'll be hidden by this catch.
Can we make this `catch` more specific to the kind of error that we want to ignore?

**Review by mitchhentgesspotify — CHANGES_REQUESTED** [review]
Nice work! This is a lot of code 😳 a reviewing marathon

File: packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts (hunk 1 above)
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts:72]: Yeah this  make sense to refactor a bit. Your earlier suggestion of getting a single file instead should sounds like a good approach. I'll give that a try!

File: packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts (hunk 3 above)
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts:98]: Hmm, not sure why that catch ended up being ignored. I'll un-ignore that 👍

File: packages/cli/src/commands/onboard/discovery/analyzers/PackageJsonAnalyzer.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    const content = await readPackageJson(packageJson);
+    if (!content) {
+      return;
+    }
+
+    const name =
+      content.name && content.name !== 'root'
+        ? content.name.split('/').slice(-1)[0] // remove before slash
```
[the remark this answers is not in the document]
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/analyzers/PackageJsonAnalyzer.ts]: Fixed in https://github.com/backstage/backstage/pull/17624/commits/8271273c0e6537929686060a01fd349edb5b8379

File: packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts (hunk 3 above)
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts]: https://github.com/backstage/backstage/pull/17624/commits/6017ae1278e6e93697980f70548a75a5cbbb079f

File: packages/cli/src/commands/onboard/discovery/providers/github/graphql.ts (hunk 2 above)
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/providers/github/graphql.ts]: https://github.com/backstage/backstage/pull/17624/commits/73965ce76ab59987ec8c56c90200f6504c769b04

File: packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts (hunk 1 above)
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts]: https://github.com/backstage/backstage/pull/17624/commits/94dcf05da54fc197c1a155dc8bd12a9dad17b50a

File: packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts [hunk 5]
```diff
@@ -47,12 +46,12 @@ export class GithubRepository implements Repository {
...
   get description(): string | undefined {
-    return this.#repo.description;
+    return this.#repo.description ?? undefined;
   }
 
   async file(filename: string): Promise<RepositoryFile | undefined> {
     const content = await this.#getFileContent(filename);
-    if (!content || content.isBinary) {
+    if (!content || content.isBinary || !content.text) {
       return undefined;
```
[the remark this answers is not in the document]
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts]: https://github.com/backstage/backstage/pull/17624/commits/bb1b1769444512f9fcf6d239f097ceaa8a3de2a9

File: packages/cli/src/commands/onboard/discovery/analyzers/PackageJsonAnalyzer.ts [hunk 6]
```diff
@@ -101,6 +98,16 @@ const packageSchema = z.object({
   name: z.string().optional(),
 });
 
+/**
+ * Makes sure that a name retrieved from a package.json file
+ * is reasonable and conforms to the catalog naming format.
+ */
+function sanitizeName(name?: string) {
+  return name && name !== 'root'
+    ? name.replace(/[^a-z0-9A-Z]/g, '_').substring(0, 62)
```
[the remark this answers is not in the document]
Comment (marcuseide) [reply packages/cli/src/commands/onboard/discovery/analyzers/PackageJsonAnalyzer.ts:111]: https://github.com/backstage/backstage/pull/17624/commits/bb1b1769444512f9fcf6d239f097ceaa8a3de2a9

File: packages/integration/src/gitlab/SingleInstanceGitlabCredentialsProvider.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+export class SingleInstanceGitlabCredentialsProvider
+  implements GitlabCredentialsProvider
+{
+  static create: (
+    config: GitLabIntegrationConfig,
+  ) => GitlabCredentialsProvider = config => {
+    return new SingleInstanceGitlabCredentialsProvider(config.token);
+  };
+
+  constructor(private readonly token?: string) {}
```
Comment (freben) [inline packages/integration/src/gitlab/SingleInstanceGitlabCredentialsProvider.ts:29]: private?

File: packages/integration/src/gitlab/SingleInstanceGitlabCredentialsProvider.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+  async getCredentials(): Promise<GitlabCredentials> {
+    return {
+      headers: this.token
+        ? { Authorization: `Bearer ${this.token}` }
+        : undefined,
+      token: this.token,
```
Comment (freben) [inline packages/integration/src/gitlab/SingleInstanceGitlabCredentialsProvider.ts:36]: Might be worth it to be thorough here, and just do `if (!this.token) { return {}; }` at the top so we don't get keys with undefined in them unnecessarily. Doesn't make any practical difference in al honesty, just a bit of extra ✨

File: .changeset/two-mayflies-tell.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/integration': patch
```
Comment (freben) [inline .changeset/two-mayflies-tell.md:2]: This is a 1.x package so a new non-breaking feature should be `minor`.

Should probably make a separate changeset for this one too! Saying like "Added X and Y" which gives sufficient context.

File: packages/cli/src/commands/onboard/command.ts [hunk 10]
```diff
@@ -50,7 +60,11 @@ export async function command(): Promise<void> {
     await integrations(providerInfo);
   }
 
-  if (!shouldSetupAuth && !shouldSetupScaffolder) {
+  if (shouldDiscoverEntities) {
+    await discover(providerInfo);
+  }
+
+  if (!shouldSetupAuth && !shouldSetupScaffolder && !shouldDiscoverEntities) {
```
Comment (freben) [inline packages/cli/src/commands/onboard/command.ts:67]: Starting to feel that maybe this check could have been at the top, with an early return. No rush to change that though.

File: packages/cli/src/commands/onboard/discovery/index.ts [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  provider: string;
+  answers: GitHubAnswers | GitLabAnswers;
+}) {
+  Task.log(`
+    Would you like to scan for - and create - Software Catalog entities?
+
+    You will need to select which SCM (Source Code Management) provider you are using, 
+    and then which repository or organization you want to scan.
+
+    This will generate a new file in the root of your project containing discovered entities,
```
Comment (freben) [inline packages/cli/src/commands/onboard/discovery/index.ts:42]: possible risk of confusion of the word "project" here - can it be read as expecting a file to be created in the remote repo being scanned? Not sure.

File: packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      url,
+    });
+
+    if (credentials.headers) {
+      return credentials.headers;
+    } else if (credentials.token) {
+      return { authorization: `Bearer ${credentials.token}` };
+    }
+
+    if (this.#envToken) {
```
Comment (freben) [inline packages/cli/src/commands/onboard/discovery/providers/gitlab/GitlabDiscoveryProvider.ts:98]: could be argued that if you went through the trouble of supplying an env token by hand, it should "win" over the other alternative above

File: packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    const content = await this.#getFileContent(filename);
+    if (!content || content.isBinary || !content.text) {
+      return undefined;
+    }
+
+    return new GithubFile(filename, content.text ?? '');
+  }
+
+  async #getFileContent(filename: string) {
+    const query = `query RepoFiles($owner: String!, $name: String!, $expr: String!) {
```
Comment (benjdlambert) [inline packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts:62]: Does this also need to be paginated?

**Review by benjdlambert — COMMENTED** [review]
Looking pretty good to me, just some higher level things I thought of.

**Comment by marcuseide** [comment]
@benjdlambert Would it be possible to get this in, and then iterate on it for your comments?

**Review by freben — APPROVED** [review]
I'm happy!
